### DISCUSSION TITLE
Support for Custom Actions on Browse/Detail

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#3047] [Feature Requests]
It's come up quite a bit that admins want to add a way to run X custom code from the browse or detail page. While creating custom layouts is one option, it's not the most efficient way to add an action button or sidebar component.

I don't know about opening up action buttons, as that is a very "holy" area in the App, and we might want to reserve it for official use. But it would be ideal if there was a way to register modular Info Sidebar Components to specific Browse/Detail pages — perhaps scoped by filters or collection. Some potential config options would then be:

* `Name` — Shown in the component header title
* `Icon` — Shown in the component header title
* `Scope` — Some sort of filter or collection allow-list that decides where the component shows
* `Code` — This is not an option, just the file that renders inside the info sidebar component when active

**Comment by mdocter** [comment]
This is exactly what I was about to request Ben. Imho the app is now marketed too much as an 'admin' app to manage the headless CMS, but it has the potential to be a full blown and extendable client. Why would I write my own JS front-end/PWA/etc, if I can customize / extend the Directus app in such a way that it has everything most apps need.
But in that case it requires features like this. 

Webiny (webiny.com) accomplishes this app extensibility by using a smart plugin system: 

   https://docs.webiny.com/docs/webiny-apps/admin/development/plugins-reference/app

Plugins are registered at app start-up in an in-memory DI container and can be retrieved at runtime/on-demand.

I myself are active in the Microsoft PowerApps / Dynamics 365 world and there a JavaScript SDK is provided by Microsoft:

   https://docs.microsoft.com/en-us/powerapps/developer/model-driven-apps/clientapi/reference

Custom JavaScript files (=webresources) are even loaded in JS sandboxes for extra 'safety', so they can't access the full DOM and Window object. Scripts only have access to certain properties and the SDK provided methods and objects.   

Those app action buttons require also some context data (or props) as input that contain for example the selected record id's (when placed above layouts) or the current record (when placed on the Item Detail page).

Just some ideas...

**Comment by rijkvanzanten** [comment]
This now exists in the form of Flows manual triggers!
